packages/jupyter-ai/jupyter_ai/context_providers/base.py

```diff
@@ -0,0 +1,239 @@
+        # same as BaseChatHandler.output_dir
+        if self.preferred_dir and os.path.exists(self.preferred_dir):
+            return self.preferred_dir
+        else:
+            return self.root_dir
+
+    def get_llm(self):
+        lm_provider = self.config_manager.lm_provider
+        lm_provider_params = self.config_manager.lm_provider_params
+
+        curr_lm_id = (
+            f'{self.llm.id}:{lm_provider_params["model_id"]}' if self.llm else None
+        )
+        next_lm_id = (
+            f'{lm_provider.id}:{lm_provider_params["model_id"]}'
+            if lm_provider
+            else None
+        )
+
+        if not lm_provider or not lm_provider_params:
+            return None
+
+        if curr_lm_id != next_lm_id:
+            model_parameters = self.model_parameters.get(
+                f"{lm_provider.id}:{lm_provider_params['model_id']}", {}
+            )
+            unified_parameters = {
+                "verbose": True,
+                **lm_provider_params,
+                **model_parameters,
+            }
+            llm = lm_provider(**unified_parameters)
+            self.llm = llm
+        return self.llm
+
+
+class ContextCommand(BaseModel):
+    cmd: str
+
+    @property
+    def id(self) -> str:
+        return self.cmd.partition(":")[0]
+
+    @property
+    def arg(self) -> Optional[str]:
+        if ":" not in self.cmd:
+            return None
+        return self.cmd.partition(":")[2].strip("'\"").replace("\\ ", " ")
+
+    def __str__(self) -> str:
+        return self.cmd
+
+    def __hash__(self) -> int:
+        return hash(self.cmd)
+
+
+class BaseCommandContextProvider(_BaseContextProvider):
+    id_prefix: ClassVar[str] = "@"
+    """Prefix symbol for command. Generally should not be overridden."""
+
+    # Configuration
+    requires_arg: ClassVar[bool] = False
+    """Whether command has an argument. E.g. '@file:<filepath>'."""
+    remove_from_prompt: ClassVar[bool] = False
+    """Whether the command should be removed from prompt when passing to LLM."""
+    only_start: ClassVar[bool] = False
+    """Whether to command can only be inserted at the start of the prompt."""
+
+    @property
+    def command_id(self) -> str:
+        return self.id_prefix + self.id
+
+    @property
+    def pattern(self) -> str:
+        # arg pattern allows for arguments between quotes or spaces with escape character ('\ ')
+        return (
+            rf"(?<![^\s.])@{self.id}:(?:'[^']+'|\"[^\"]+\"|[^\s\\]+(?:\\ [^\s\\]*)*)"
+            if self.requires_arg
+            else rf"(?<![^\s.])@{self.id}(?![^\s.])"
+        )
+
+    async def make_context_prompt(self, message: HumanChatMessage) -> str:
+        """Returns a context prompt for all commands of the context provider
+        command.
+        """
+        commands = self._find_commands(message.prompt)
+        if not commands:
+            return ""
+        return await self._make_context_prompt(message, commands)
```

Comment:
sent you an email to your david@qiu.dev address. includes some security related stuff so i wasn't sure if it would be better kept out of a public forum.

Reply:
you can ignore the suggestions in the email i sent as i found a simple way to meet the requirement above. however the general points i made are still valid and may come into play when thinking about more general yet extendible context providers.

Reply:
Thanks, awesome work!